src/metrics/rules/active_ruleset_bench_test.go

```diff
@@ -0,0 +1,278 @@
+package rules
+
+import (
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/metrics/aggregation"
+	"github.com/m3db/m3/src/metrics/filters"
+	"github.com/m3db/m3/src/metrics/metric"
+	metricid "github.com/m3db/m3/src/metrics/metric/id"
+	"github.com/m3db/m3/src/metrics/metric/id/m3"
+	"github.com/m3db/m3/src/metrics/pipeline"
+	"github.com/m3db/m3/src/metrics/policy"
+	xtime "github.com/m3db/m3/src/x/time"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var benchmarkCases = []struct {
+	name string
+	id   string
+}{{
+	name: "rollup ID",
+	id:   fmt.Sprintf("m3+%s+m3_rollup=true,rtagName1=val1,rtagName2=val2", newTestRuleName("rollup", 5)),
+}, {
+	name: "mapping ID",
+	id:   "m3+test+rtagName1=val1,rtagName2=val2",
+}}
+
+// Results (2024-10-18, Macbook Pro M1 Max):
+//
+// goos: darwin
+// goarch: arm64
+// pkg: github.com/m3db/m3/src/metrics/rules
+// BenchmarkActiveRuleSet_ReverseMatch
+// BenchmarkActiveRuleSet_ReverseMatch/rollup_ID
+// BenchmarkActiveRuleSet_ReverseMatch/rollup_ID-10         	      21	  52730040 ns/op	43040917 B/op	  360008 allocs/op
+// BenchmarkActiveRuleSet_ReverseMatch/mapping_ID
+// BenchmarkActiveRuleSet_ReverseMatch/mapping_ID-10        	       2	 533871292 ns/op	1089136572 B/op	 4670173 allocs/op
+// PASS
+func BenchmarkActiveRuleSet_ReverseMatch(b *testing.B) {
```

Comment:
this is great. i think you should add a few things to see how this behaviour changes as the code is used in different ways:

1/ add a case which runs thru the code with different IDs instead of just one - they can all have the same basic tag structure but some may have extra tags/etc
2/ you already have configs for num rules/num snapshots per rule - write a few different benchmark cases to see how the performance scales
3/ add a knob for % of rules matching the ids - e.g. 100 rules in your corpus out of which only say 10% match the given input ids.

Reply:
fine by me to do in a followup. but i think you should change the order of next work: add more cases to your benchmark that represent production load and then make your savings changes. 

you want more data to know how effective your changes are.